pkg/bitwarden/cli.go

```diff
@@ -0,0 +1,153 @@
+package bitwarden
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type bitwardenCliClient struct {
+	username string
+	password string
+	sync.Mutex
+	session    string
+	savedItems []Item
+	run        func(cmd string, args ...string) ([]byte, error)
+}
+
+func newBitwardenCliClient(username, password string) (Client, error) {
+	return newBitwardenCliClientWithRun(username, password, func(cmd string, args ...string) (bytes []byte, err error) {
+		return exec.Command(cmd, args...).Output()
+	})
+}
+
+func newBitwardenCliClientWithRun(username, password string, run func(cmd string, args ...string) (bytes []byte, err error)) (Client, error) {
+	client := bitwardenCliClient{
+		username: username,
+		password: password,
+		run:      run,
+	}
+	return &client, client.loginAndListItems()
+}
+
+type bwLoginResponse struct {
+	Success bool `json:"success"`
+	Data    struct {
+		Raw string `json:"raw"`
+	} `json:"data"`
+}
+
+func (c *bitwardenCliClient) loginAndListItems() error {
+	c.Lock()
+	defer c.Unlock()
+	output, err := c.runBWCmd(bwLoginCmd, c.username, c.password, "--response")
+	if err != nil {
+		return err
+	}
+	r := bwLoginResponse{}
+	if err := json.Unmarshal(output, &r); err != nil {
+		//TODO need to censor if we plan this on a job
+		return err
+	}
+	if r.Success {
+		raw := r.Data.Raw
+		if raw != "" {
+			c.session = raw
+			if len(c.savedItems) > 0 {
+				return nil
+			}
+			var items []Item
+			out, err := c.runBWCmd(bwListCmd, "items")
+			if err != nil {
+				return err
+			}
+			err = json.Unmarshal(out, &items)
+			if err != nil {
+				return err
+			}
+			c.savedItems = items
+			return nil
+		}
+		// should never happen
+		return fmt.Errorf("bw-login succeeded with empty '.data.raw'")
+	}
+	// should never happen
+	return fmt.Errorf("bw-login failed without run's error")
+}
+
+func (c *bitwardenCliClient) GetItem(itemName string) (ItemGetter, error) {
+	var itemGetter cliItemGetter
+	for _, i := range c.savedItems {
+		if i.Name == itemName {
+			itemGetter.item = i
+			itemGetter.client = c
+			return &itemGetter, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to find item with name %s", itemName)
```

Comment:
What do I see if I `GetItem().GetField()` and `GetItem()` returns `nil`?

Reply:
I imagine a nil pointer panic?

Reply:
Back to 

```
GetFieldOnItem(itemName, fieldName string) (string, error)
GetAttachmentOnItem(itemName, attachmentName string) ([]byte, error)
```